RFC-0001-torch-function-for-methods.md

```diff
@@ -0,0 +1,177 @@
+4. For one argument of each type, `__torch_function__` is called. The first such function or method to return something other than `NotImplemented` will be the final result. All exceptions will be propagated upward.
+5. If all `__torch_function__` implementations return `NotImplemented`, a `TypeError` is raised with an appropriate error message.
+
+In practice, for most PyTorch functions, the list of tensor-likes is already available and the dispatcher doesn't need to be called. Additionally, while equivalent to the code above, if the `Tensor`-likes are all `Tensor`s or classes for which `__torch_function__ is torch.Tensor.__torch_function__`, the internal implementation is called immediately. This is done as a performance optimisation to avoid overhead for concrete `Tensor` objects.
+
+To access super, one would do the following:
+```python
+class SubTensor(torch.Tensor):
+    def __torch_tensor__(self, func, types, args, kwargs):
+        # Pre-processing here
+        val = super().__torch_function__(
+            func,
+            tuple(
+                t for t in types if not issubclass(t, SubTensor)
+            ),
+            args,
+            kwargs
+        )
+        # Post processing here
+```
+
+This way `__torch_function__` knows the list of types to dispatch to, and it will _not_ dispatch to `SubTensor` anymore in this example.
+
+We will also recommend that all `Tensor` subclasses make their own methods go through `__torch_function__` via a decorator `@torch_function_dispatch`. This decorator was added and then removed for performance reasons, however it will be added back to allow external libraries to interface with the protocol. It will take a single argument: a dispatcher, i.e. a callable that returns an iterable of all the "duck-Tensors", or possible candidates for classes that may implement `__torch_function__`.
+
+If a library forgets to add the aforementioned decorator, then the method will no longer dispatch at all to any form of `__torch_function__`. In other words, it will lose support for the protocol.
+
+However, this will come with a disclaimer: They _must_ accept that their methods are subject to the same processing as any other `torch.Tensor` methods, namely, that all the processing _will necessarily go through `__torch_function__`, even if through superclasses first_. Specifically, processing may pass through a superclass's `__torch_function__` implementation before coming back to a subclass's internal implementation. A workaround would be to define a `__torch_function__` on the same class that directly calls the internal implementation, or follows the same pattern as the example `__torch_function__` defined on `torch.Tensor`. This will allow subclasses to behave exactly like `torch.Tensor`, while still allowing access to method overriding via the `__torch_function__` protocol.
```

Comment:
This paragraph needs a rewrite. It's unclear whether it's important (it's just a disclaimer), it's unclear what a user should actually do (does "must accept ..." imply they need to add some code or not?), and what the penalty is.